scipy/optimize/cython_optimize/__init__.py

```diff
@@ -0,0 +1,138 @@
+"""
+Cython Optimize Zeros API
+=========================
+The following four zeros functions can be accessed directly from Cython:
+
+- ``bisect``
+- ``ridder``
+- ``brenth``
+- ``brentq``
+
+Import the module into Cython as follows::
+
+    from scipy.optimize cimport cython_optimize
+
+The Cython Optimize Zeros API could be used to "vectorize" a root finder given
+a single callback function and an array of input conditions using a for-loop or
+``prange`` in Cython.
+
+
+Callback Signature
+------------------
+The zeros functions in this module use a callback that takes a double with the
+scalar independent variable as the 1st argument and a user defined ``struct``
+with any extra parameters as the 2nd argument. ::
+
+    double (*callback_type)(double, void*)
+
+Import the module into Cython as follows::
+
+    from scipy.optimize.cython_optimize cimport zeros
+
+
+Examples
+--------
+Usage of ``scipy.optimize.cython_optimize`` requires Cython to write callbacks
+that are compiled into C. For more information on compiling Cython see the
+`Cython Documentation <http://docs.cython.org/en/latest/index.html>`_.
+
+These are the basic steps:
+
+1. Create a Cython ``.pyx`` file
+2. Import the ``cython_optimize.zeros`` module
+3. Write the callback function, and call the selected zeros function passing
+   the callback, any extra arguments, and the other solver parameters ::
+
+       import math
+       from scipy.optimize.cython_optimize cimport zeros
+
+       ARGS = {'C0': 1.0, 'C1': 0.7}  # a dictionary of extra arguments
+       XLO, XHI = 0.5, 1.0  # lower and upper search boundaries
+       XTOL, RTOL, MITR = 1e-3, 1e-3, 10  # other solver parameters
+
+       # user defined struct for extra parameters
+       ctypedef struct test_params:
+           double C0
+           double C1
+
+
+       # user defined callback
+       cdef double f(double x, void *args):
+           cdef test_params *myargs = <test_params *> args
+           return myargs.C0 - math.exp(-(x - myargs.C1))
+
+
+       # Cython wrapper function
+       cdef double brentq_wrapper_example(dict args, double xa, double xb,
+                                          double xtol, double rtol, int mitr):
+           # Cython automatically casts dictionary to struct
+           cdef test_params myargs = args  
+           return zeros.brentq(
+               f, xa, xb, <test_params *> &myargs, xtol, rtol, mitr, NULL)
+
+
+       # Python function
+       def brentq_example(args=ARGS, xa=XLO, xb=XHI, xtol=XTOL, rtol=RTOL,
+                          mitr=MITR):
+           '''Calls Cython wrapper from Python.'''
+           return brentq_wrapper_example(args, xa, xb, xtol, rtol, mitr)
+
+
+       x = brentq_example()
```

Comment:
this is probably a little confusing; the rest of this example lives in the `.pyx` file, while this does't

Reply:
Is this better? I've moved the Python portion after step 4, added an import statement, and the following text:

>then in Python  you can import and run the example

Reply:
In actuality I run this example in a Jupyter with the Cython magic `%cython` and it just works.

Reply:
lgtm